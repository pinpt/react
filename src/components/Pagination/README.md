### A component to render pagination controls on a page

[Full Docs](https://react.preview.pinpoint.com/?path=/docs/components-pagination) | [Github](https://github.com/pinpt/react/tree/master/src/components/Pagination)

#### Theming

The parent className for styling this component is `.Pagination`

<details>
	<summary>Base Theme Styles</summary>

```css
.paginationWrapper {
	@apply pb-10 md:pb-14 flex-grow;
}

.paginationWrapper .Pagination {
	@apply md:ml-[calc(25%+1rem)];
}

.Pagination {
	@apply flex justify-between items-center space-x-10;
	color: var(--page-highlight-color);
}

.Pagination.forwardOnly {
	@apply justify-end;
}

.Pagination .back,
.Pagination .forward {
	@apply min-w-0 flex items-center cursor-pointer;
	flex: 1 1 50%;
}

.Pagination .forward {
	@apply justify-end;
}

.Pagination .forward .icon {
	@apply ml-2 flex-shrink-0;
}

.Pagination .back .icon {
	@apply mr-2 flex-shrink-0;
}

.Pagination .back .text,
.Pagination .forward .text {
	@apply truncate;
}
```

</details>
