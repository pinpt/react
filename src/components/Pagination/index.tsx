export interface IPaginationProps {
	className?: string;
	goForward?: () => void;
	goBack?: () => void;
}

const Pagination = (props: IPaginationProps) => {
	const { className = '', goForward, goBack } = props;

	return (
		<div className={`Pinpoint Pagination ${className}`}>
			{goBack && (
				<div className="back" onClick={goBack}>
					<svg
						className="icon"
						width={15.75}
						height={18}
						aria-hidden="true"
						focusable="false"
						data-prefix="fas"
						data-icon="arrow-left"
						role="img"
						xmlns="http://www.w3.org/2000/svg"
						viewBox="0 0 448 512"
					>
						<path
							fill="currentColor"
							d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
						></path>
					</svg>
					Back
				</div>
			)}
			{goForward && (
				<div className="forward" onClick={goForward}>
					Next
					<svg
						className="icon"
						width={15.75}
						height={18}
						aria-hidden="true"
						focusable="false"
						data-prefix="fas"
						data-icon="arrow-right"
						role="img"
						xmlns="http://www.w3.org/2000/svg"
						viewBox="0 0 448 512"
					>
						<path
							fill="currentColor"
							d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
						></path>
					</svg>
				</div>
			)}
		</div>
	);
};

export default Pagination;
