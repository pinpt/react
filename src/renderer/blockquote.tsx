import { registerNode, recurseIntoChildren, NodeProps } from './register';

const BlockQuote = ({ node }: NodeProps) => {
	return <blockquote>{recurseIntoChildren(node)}</blockquote>;
};

registerNode('blockquote', (node) => <BlockQuote node={node} />);
