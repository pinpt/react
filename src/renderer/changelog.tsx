import { ReactYouTubeLite as Youtube } from 'react-youtube-lite';
import { getFileUrl } from '../lib/file_metadata';
import { slugifyContent } from '../lib/string';
import { ChangelogCoverMedia, ChangelogMediaType } from '../lib/types';
import { Content as Doc } from './';

const ImageMedia = ({ src, title = '' }: { src: string; title?: string }) => {
	return (
		<div className="image">
			<img src={src} alt={title} />
		</div>
	);
};

const VideoMedia = ({ src }: { src: string }) => {
	return (
		<div className="video">
			<video controls src={src} />
		</div>
	);
};

const YoutubeMediaSSR = ({ id, poster = 'hqdefault' }: { id: string; poster?: string }) => {
	// the component isn't SSR friendly so this is basically used to sort of mimic it for generated code
	const posterUrl = `https://i.ytimg.com/vi/${id}/${poster}.jpg`;
	return (
		<>
			{poster && <link rel="preload" href={posterUrl} as="image" />}
			<div
				className="youtube ryt-lite embed-responsive aspect-ratio-16/9"
				style={{ backgroundImage: `url(${posterUrl})` }}
			>
				<div className="lty-playbtn" data-url={`https://www.youtube-nocookie.com/embed/${id}?autoplay=1`} />
			</div>
		</>
	);
};

const YoutubeMedia = ({
	id,
	metadata,
	staticMode,
}: {
	id: string;
	metadata?: Record<string, any>;
	staticMode?: true;
}) => {
	return (
		<div className="youtube">
			{!staticMode ? (
				<Youtube url={`https://www.youtube.com/watch?v=${id}`} poster={metadata?.poster} />
			) : (
				<YoutubeMediaSSR id={id} poster={metadata?.poster} />
			)}
		</div>
	);
};

const Media = ({ media, title, staticMode }: { media?: ChangelogCoverMedia; title?: string; staticMode?: true }) => {
	if (!media) {
		return <></>;
	}
	let content: React.ReactElement;
	switch (media.type) {
		case ChangelogMediaType.None: {
			return <></>;
		}
		case ChangelogMediaType.Image: {
			content = <ImageMedia src={getFileUrl(media.value)} title={title} />;
			break;
		}
		case ChangelogMediaType.Video: {
			content = <VideoMedia src={getFileUrl(media.value)} />;
			break;
		}
		case ChangelogMediaType.Youtube: {
			content = <YoutubeMedia id={media.value} metadata={media.metadata} staticMode={staticMode} />;
			break;
		}
	}
	return <div className="media-container">{content}</div>;
};

interface ChangelogProps {
	id: string;
	title?: string;
	content?: any;
	coverMedia?: ChangelogCoverMedia;
	limit?: number;
	divider?: boolean;
	staticMode?: true;
}

const ChangelogLog = (props: ChangelogProps) => {
	return (
		<article className="changelog notebook-editor read-only">
			{props.coverMedia && (
				<section className="covermedia">
					<Media media={props.coverMedia} title={props.title} staticMode={props.staticMode} />
				</section>
			)}
			<section className="ProseMirror">
				<Doc node={props.content} limit={props.limit} />
				{props.limit && props.content.content?.length > props.limit && (
					<div className="continue">
						<a href={slugifyContent(props.id, props.title)}>Continue Reading â†’</a>
					</div>
				)}
			</section>
			{props.divider && <hr className="divider" />}
		</article>
	);
};

export default ChangelogLog;
