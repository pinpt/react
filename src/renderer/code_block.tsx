import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { twilight } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { NodeProps, registerNode } from './register';

const CodeBlock = ({ node }: NodeProps) => {
	return (
		<SyntaxHighlighter
			language={node.attrs.language}
			style={twilight}
			customStyle={{
				background: 'black',
				fontFamily: 'ui-monospace',
				borderWidth: '0px',
				borderRadius: '.25em',
			}}
		>
			{node.content?.[0].text ?? ''}
		</SyntaxHighlighter>
	);
};

registerNode('code_block', (node) => <CodeBlock node={node} />);
