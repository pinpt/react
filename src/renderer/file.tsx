import {
	faDownload, faFile, faFileAlt, faFileArchive, faFileAudio, faFileCode, faFileExcel, faFilePdf,
	faFilePowerpoint, faFileVideo, faFileWord
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { NodeProps, registerNode } from './register';

export const iconForType = (type = '') => {
	if (/pdf/.test(type)) {
		return faFilePdf;
	}
	if (/zip|tar/.test(type)) {
		return faFileArchive;
	}
	if (/video/.test(type)) {
		return faFileVideo;
	}
	if (/audio/.test(type)) {
		return faFileAudio;
	}
	if (/excel|spreadsheet/.test(type)) {
		return faFileExcel;
	}
	if (/powerpoint/.test(type)) {
		return faFilePowerpoint;
	}
	if (/word/.test(type)) {
		return faFileWord;
	}
	if (/css|csv|graphql|html|java|json|php|python|typescript|xml/.test(type)) {
		return faFileCode;
	}
	if (/text/.test(type)) {
		return faFileAlt;
	}
	return faFile;
};

export const formattedSize = (size = 0) => {
	const kB = size / 1000;
	if (kB < 10) {
		return `${Math.round(kB * 10) / 10} KB`;
	}
	if (kB < 1000) {
		return `${Math.round(kB)} KB`;
	}
	const MB = kB / 1000;
	return `${Math.round(MB * 10) / 10} MB`;
};

const File = ({ node }: NodeProps) => {
	const { description, filename, publicUrl, size, type } = node.attrs;

	const onDownload = () => {
		if (publicUrl) {
			const u = new URL(publicUrl);
			u.searchParams.set('filename', filename);
			u.searchParams.set('disposition', 'attachment');
			const a = document.createElement('a');
			a.style.display = 'none';
			a.href = u.toString();
			// the filename you want
			a.download = filename;
			document.body.appendChild(a);
			a.click();
			a.remove();
		}
	};

	return (
		<div className="file">
			<div className="container">
				<FontAwesomeIcon icon={iconForType(type)} />
				<div className="details">
					{publicUrl ? (
						<a href={publicUrl} target="_blank" rel="noopener noreferrer" className="filename">
							{filename}
						</a>
					) : (
						<div className="filename">{filename}</div>
					)}
					{description && <div className="description">{description}</div>}
					<div className="size">{formattedSize(size)}</div>
				</div>
				<div className="options">
					{publicUrl && (
						<div className="option group" onClick={onDownload}>
							<FontAwesomeIcon icon={faDownload} />
							<span>Download</span>
						</div>
					)}
				</div>
			</div>
		</div>
	);
};

registerNode('file', (node) => <File node={node} />);
