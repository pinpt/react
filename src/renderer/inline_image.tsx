import { extractImageMetadataFromFileID } from '../lib/file_metadata';
import { NodeProps, registerNode } from './register';

const InlineImage = ({ node }: NodeProps) => {
	const { size } = extractImageMetadataFromFileID(node.attrs?.src);
	const width = size?.width;
	const height = size?.height;

	const image = (
		<img src={node.attrs?.src} width={width} height={height} alt={node.attrs?.alt} className="inline_image" />
	);

	return (
		<div>
			{node.attrs?.href ? (
				<a href={node.attrs.href} target="_blank" rel="noopener noreferrer">
					{image}
				</a>
			) : (
				<>{image}</>
			)}
		</div>
	);
};

registerNode('inline-image', (node) => <InlineImage node={node} />);
