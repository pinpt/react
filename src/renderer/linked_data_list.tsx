import { NodeProps, recurseIntoChildren, registerNode } from './register';

const LinkedDataList = ({ node }: NodeProps) => {
	return <div className="linked-data-list">{recurseIntoChildren(node)}</div>;
};

registerNode('linked_data_list', (node) => <LinkedDataList node={node} />);
