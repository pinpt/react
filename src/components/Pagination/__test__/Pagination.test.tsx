import renderer from 'react-test-renderer';
import Pagination from '..';

test('Test default (neither direction)', () => {
	const component = renderer.create(<Pagination />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test both directions', () => {
	const component = renderer.create(
		<Pagination goBack={() => console.log('back!')} goForward={() => console.log('forward!')} />
	);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test back only', () => {
	const component = renderer.create(<Pagination goBack={() => console.log('back!')} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test forward only', () => {
	const component = renderer.create(<Pagination goForward={() => console.log('forward!')} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
