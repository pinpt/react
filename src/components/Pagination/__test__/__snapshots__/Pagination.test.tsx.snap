// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test back only 1`] = `
<div
  className="Pinpoint Pagination "
>
  <div
    className="back"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="icon"
      data-icon="arrow-left"
      data-prefix="fas"
      focusable="false"
      height={18}
      role="img"
      viewBox="0 0 448 512"
      width={15.75}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
        fill="currentColor"
      />
    </svg>
    Back
  </div>
</div>
`;

exports[`Test both directions 1`] = `
<div
  className="Pinpoint Pagination "
>
  <div
    className="back"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="icon"
      data-icon="arrow-left"
      data-prefix="fas"
      focusable="false"
      height={18}
      role="img"
      viewBox="0 0 448 512"
      width={15.75}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
        fill="currentColor"
      />
    </svg>
    Back
  </div>
  <div
    className="forward"
    onClick={[Function]}
  >
    Next
    <svg
      aria-hidden="true"
      className="icon"
      data-icon="arrow-right"
      data-prefix="fas"
      focusable="false"
      height={18}
      role="img"
      viewBox="0 0 448 512"
      width={15.75}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`Test default (neither direction) 1`] = `
<div
  className="Pinpoint Pagination "
/>
`;

exports[`Test forward only 1`] = `
<div
  className="Pinpoint Pagination "
>
  <div
    className="forward"
    onClick={[Function]}
  >
    Next
    <svg
      aria-hidden="true"
      className="icon"
      data-icon="arrow-right"
      data-prefix="fas"
      focusable="false"
      height={18}
      role="img"
      viewBox="0 0 448 512"
      width={15.75}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
